### PR TITLE
Add edge from catch block to enclosing finally block in jitloopbodies

### PR DESCRIPTION
**Pull request by meg-gupta**
We can have BailOutOnEarlyExit which necessitates accurate flow.
This change adds a BrOnException fake flow edge from catch block's end to  finally block enclosing it.

Fixes OS#16673022

### CODE REVIEW

File: lib/Backend/FlowGraph.cpp [hunk 1]
```diff
@@ -451,18 +459,21 @@ FlowGraph::Build(void)
     }
 #endif
 
-    if (this->finallyLabelStack)
+    if (createNonExceptionFinally || this->func->IsLoopBodyInTryFinally())
```
Comment (sigatrev) [inline lib/Backend/FlowGraph.cpp]: Under what circumstances is this condition different from the condition where finallyLabelStack is created?

File: lib/Backend/FlowGraph.cpp [hunk 2]
```diff
@@ -172,18 +172,21 @@ FlowGraph::Build(void)
 
     bool assignRegionsBeforeGlobopt = this->func->HasTry() && (this->func->DoOptimizeTry() ||
         (this->func->IsSimpleJit() && this->func->hasBailout) ||
-        this->func->IsLoopBodyInTryFinally());
+        (this->func->IsLoopBodyInTryFinally() && this->func->GetJITFunctionBody()->DoJITLoopBody()));
```
Comment (rajatd) [inline lib/Backend/FlowGraph.cpp]: >DoJITLoopBody [](start = 83, length = 13)

Wouldn't this always return true if `func->IsLoopBodyInTryFinally()` returns true?

File: lib/Backend/FlowGraph.cpp [hunk 3]
```diff
@@ -172,18 +172,21 @@ FlowGraph::Build(void)
...
+
+    bool createNonExceptionFinally = this->func->HasFinally() && (this->func->DoOptimizeTry() ||
+        (this->func->IsSimpleJit() && this->func->GetJITFunctionBody()->DoJITLoopBody()));
```
Comment (rajatd) [inline lib/Backend/FlowGraph.cpp]: >DoJITLoopBody [](start = 72, length = 13)

I think using this->func->IsLoopBody() would make more sense here. Similarly at two instances below. DoJITLoopBody() also does not guarantee that we are jitting a loop body at this point.

Comment (meg-gupta) [reply lib/Backend/FlowGraph.cpp]: I realized I was using this->func->IsSimpleJit() && this->func->GetJITFunctionBody()->DoJITLoopBody(), to mean this->func->IsSimpleJit() && this->func->hasBailout

File: lib/Backend/FlowGraph.cpp (hunk 2 above)
Comment (meg-gupta) [reply lib/Backend/FlowGraph.cpp]: Right, its redundant

File: lib/Backend/FlowGraph.cpp (hunk 1 above)
Comment (meg-gupta) [reply lib/Backend/FlowGraph.cpp]: Its not, I'll change it back

**Comment by meg-gupta** [comment]
@dotnet-bot test OSX _no_jit_shared_osx_osx_test

**Comment by meg-gupta** [comment]
@dotnet-bot test OSX static_osx_osx_debug
@dotnet-bot test OSX static_osx_osx_release
@dotnet-bot test OSX static_osx_osx_release

**Pull request merged**